nixd/lib/Controller/Completion.cpp

```diff
@@ -280,6 +280,9 @@ void completeAttrPath(const Node &N, const ParentMapAnalysis &PM,
     // Construct request.
     std::string Prefix = Scope.back();
     Scope.pop_back();
+    if (!Scope.empty() && Scope[0] == "config") {
```

Comment:
This approach LGTM. However, we also use `findAttrPath` in `Definition.cpp` and `Hover.cpp` to provide go-to-definition & hover docs for options. I kindly ask if you can write a shared function for them, like `findAttrPathForOptions` pand allowing go-to-definition + hovering works inside `config =` consistently?

Reply:
How do you feel about doing it in `getNestedAttrPath` rather than introducing a new helper method?

Btw thanks for catching that doing it only in completeAttrPath lost the fix for hover ... I now also added a respective hover test :)

Reply:
> How do you feel about doing it in `getNestedAttrPath` rather than introducing a new helper method?

I think it is a good idea to introduce a new method, because `getNestedAttrPath` is not specific to NixOS module system, and I cannot infer it may omit first `config.` by just reading the function name. So I'd like to prefer a new method.

Reply:
That makes sense. While I'm at it I'd also like to improve `findAttrPath` to better deal with function calls. In order to do that I'd like to write some unit tests for it (all the test cases would be testing the same thing and I find lit tests too verbose for this).  However I found out that this isn't that easy with the current design because `findAttrPath` requires some `ParentMapAnalysis` which is obtained via the translation unit but `Controller::getTU` is private. While we could use `friend class` for such unit tests I think it would be nicer to refactor the code so that this isn't necessary? (I'm a bit out of my depth cause I have almost no experience with C++.)

Reply:
> That makes sense. While I'm at it I'd also like to improve `findAttrPath` to better deal with function calls.

Specifically, how do you plan to improve it?

Reply:
Currently if you have `foo = someFunc { ... }` then the path inside the attrset is inferred to be the empty path rather than `foo`. Yes we don't know the semantics of the function but not forgetting the prior path would be more correct and work for e.g. `lib.mkIf` and could be made to work with `lib.mkMerge` and other common nixpkgs functions. To still work in Flakes we'd need to implement some heuristics to detect whether a file is a flake or a NixOS module as well as some flake specific semantics ... haven't looked into flake parts yet.

Reply:
> more correct

I don't think correctness can be objectively measured in this context.

> and work for e.g. `lib.mkIf` and could be made to work with `lib.mkMerge` and other common nixpkgs functions.

That said, I understand the point you're making. However, I believe changes like supporting `lib.mkIf`, and potentially `lib.mkMerge` and other common Nixpkgs functions, should be handled in separate PRs.

Reply:
Sure, I'm happy to discuss the design in an issue or on Matrix. I'm fine with leaving this to separate PRs. I have introduced the `findAttrPathForOptions` you suggested and added a FIXME comment about my concern that this leads to wrong completions at the top-level of `flake.nix` files.

I ended up improving how documents are specified in tests: #679. If you like this change and merge it before this I could add another test case for a `flake.nix` file with `nixpkgs.lib.nixosSystem` where `config` will also be recognized properly with this PR because the attr path currently resets after a function (I think it would be nice to add a test case for this so that we don't stop getting this right when making future changes).